internal/auth0/action/resource_trigger_action.go

```diff
@@ -0,0 +1,205 @@
+	actionID := d.Get("action_id").(string)
+
+	currentBindings, err := api.Action.Bindings(trigger)
+	if err != nil {
+		if mErr, ok := err.(management.Error); ok && mErr.Status() == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	t := "action_id"
+
+	var updatedBindings []*management.ActionBinding
+	for _, binding := range currentBindings.Bindings {
+		if binding.Action.ID == &actionID {
+			d.SetId(trigger + "::" + actionID)
+			return nil
+		}
+		updatedBindings = append(updatedBindings, &management.ActionBinding{
+			Ref: &management.ActionBindingReference{
+				Type:  &t,
+				Value: binding.Action.ID,
+			},
+		})
+	}
+
+	action, err := api.Action.Read(actionID)
+	if err != nil {
+		if mErr, ok := err.(management.Error); ok && mErr.Status() == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	updatedBindings = append(updatedBindings, &management.ActionBinding{
+		Ref: &management.ActionBindingReference{
+			Type:  &t,
+			Value: &actionID,
+		},
+		DisplayName: action.Name,
+	})
+
+	if err := api.Action.UpdateBindings(trigger, updatedBindings); err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(trigger + "::" + actionID)
+	return nil
+}
+
+func readTriggerAction(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	api := m.(*config.Config).GetAPI()
+
+	trigger := d.Get("trigger").(string)
+	actionID := d.Get("action_id").(string)
+
+	triggerBindings, err := api.Action.Bindings(trigger)
+	if err != nil {
+		if mErr, ok := err.(management.Error); ok && mErr.Status() == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	for _, binding := range triggerBindings.Bindings {
+		if binding.Action.ID == &actionID {
+			d.SetId(trigger + "::" + actionID)
+			return nil
+		}
+	}
+
+	d.SetId(trigger + "::" + actionID)
+	return nil
+}
+
+func deleteTriggerAction(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	api := m.(*config.Config).GetAPI()
+
+	trigger := d.Get("trigger").(string)
+	actionID := d.Get("action_id").(string)
+
+	triggerBindings, err := api.Action.Bindings(trigger)
+	if err != nil {
+		if mErr, ok := err.(management.Error); ok && mErr.Status() == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	var updatedBindings []*management.ActionBinding
+	for _, binding := range triggerBindings.Bindings {
+		if binding.Action.GetID() == actionID {
+			continue
+		}
+
+		t := "action_id"
+		id := binding.Action.GetID()
+		displayName := binding.GetDisplayName()
+		updatedBindings = append(updatedBindings, &management.ActionBinding{
+			Ref: &management.ActionBindingReference{
+				Type:  &t,
+				Value: &id,
+			},
+			DisplayName: &displayName,
+		})
+	}
+
+	if err := api.Action.UpdateBindings(trigger, updatedBindings); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
```

Comment:
Setting ID to empty not needed because the TF SDK does it for us during destroy. See https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/resource.go#L814 and https://github.com/hashicorp/terraform-provider-aws/blob/main/internal/service/ec2/ec2_instance.go for reference.

Reply:
👍🏻 Good to know, we can probably get rid of a lot of `data.SetId("")` in destroy funcs.